### PR TITLE
feat(extension): Account Abstraction

### PR DESCRIPTION
**Pull request by jxom**
Adds support for an ERC-4337 Account Abstraction extension.

[See Docs](https://viem-git-account-abstraction-wevm.vercel.app/account-abstraction)

### CODE REVIEW

File: src/account-abstraction/actions/bundler/estimateUserOperationGas.ts [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+          /** Retrieves paymaster-related User Operation properties to be used for sending the User Operation. */
+          getPaymasterData?: PaymasterActions['getPaymasterData'] | undefined
+          /** Retrieves paymaster-related User Operation properties to be used for gas estimation. */
+          getPaymasterStubData?:
+            | PaymasterActions['getPaymasterStubData']
+            | undefined
+        }
+      | undefined
+    /** Paymaster context to pass to `getPaymasterData` and `getPaymasterStubData` calls. */
+    paymasterContext?: unknown
```
Comment (tmm) [inline src/account-abstraction/actions/bundler/estimateUserOperationGas.ts:63]: `undefined` gets absorbed by `unknown`, but better to have it than no imo

```suggestion
    paymasterContext?: unknown | undefined
```

File: src/account-abstraction/actions/bundler/getSupportedEntryPoints.ts [hunk 2]
```diff
@@ -0,0 +1,32 @@
+import type { Address } from 'abitype'
+import type { Client } from '../../../clients/createClient.js'
+import type { Transport } from '../../../clients/transports/createTransport.js'
+import type { ErrorType } from '../../../errors/utils.js'
+
+export type GetSupportedEntryPointsReturnType = readonly Address[]
+export type GetSupportedEntryPointsErrorType = ErrorType
```
Comment (tmm) [inline src/account-abstraction/actions/bundler/getSupportedEntryPoints.ts:7]:
```suggestion
export type GetSupportedEntryPointsErrorType = RequestErrorType | ErrorType
```

File: src/account-abstraction/actions/bundler/prepareUserOperation.ts [hunk 3]
```diff
@@ -0,0 +1,612 @@
...
+          /** Retrieves paymaster-related User Operation properties to be used for sending the User Operation. */
+          getPaymasterData?: PaymasterActions['getPaymasterData'] | undefined
+          /** Retrieves paymaster-related User Operation properties to be used for gas estimation. */
+          getPaymasterStubData?:
+            | PaymasterActions['getPaymasterStubData']
+            | undefined
+        }
+      | undefined
+    /** Paymaster context to pass to `getPaymasterData` and `getPaymasterStubData` calls. */
+    paymasterContext?: unknown
```
Comment (tmm) [inline src/account-abstraction/actions/bundler/prepareUserOperation.ts:154]:
```suggestion
    paymasterContext?: unknown | undefined
```

File: src/account-abstraction/actions/bundler/sendUserOperation.ts [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+      | undefined
+    /** Paymaster context to pass to `getPaymasterData` and `getPaymasterStubData` calls. */
+    paymasterContext?: unknown
+  }
+> &
+  GetSmartAccountParameter<account, accountOverride>
+
+export type SendUserOperationReturnType = Hex
+
+export type SendUserOperationErrorType = ErrorType
```
Comment (tmm) [inline src/account-abstraction/actions/bundler/sendUserOperation.ts:68]:
```suggestion
export type SendUserOperationErrorType = RequestErrorType | ErrorType
```

File: src/account-abstraction/actions/paymaster/getPaymasterData.ts [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+ *
+ * const userOperation = { ... }
+ *
+ * const values = await getPaymasterData(paymasterClient, {
+ *   chainId: 1,
+ *   entryPointAddress: '0x...',
+ *   ...userOperation,
+ * })
+ */
+export async function getPaymasterData(
```
Comment (tmm) [inline src/account-abstraction/actions/paymaster/getPaymasterData.ts:97]: Should we add an error type for this?

File: src/account-abstraction/actions/paymaster/getPaymasterStubData.ts [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+ *
+ * const userOperation = { ... }
+ *
+ * const values = await getPaymasterStubData(paymasterClient, {
+ *   chainId: 1,
+ *   entryPointAddress: '0x...',
+ *   ...userOperation,
+ * })
+ */
+export async function getPaymasterStubData(
```
Comment (tmm) [inline src/account-abstraction/actions/paymaster/getPaymasterStubData.ts:100]: Should we add an error type for this?

File: src/account-abstraction/constants/abis.ts
Comment (tmm) [inline src/account-abstraction/constants/abis.ts:1]: Cool if I remove `internalType` from all ABIs?

**Review by tmm — APPROVED** [review]

**Comment by Silur** [comment]
the examples provided (https://viem.sh/account-abstraction/accounts/webauthn/createWebAuthnCredential) do no work neither in firefox nor in chrome:

```
import { 
  createWebAuthnCredential, 
  toWebAuthnAccount 
} from 'viem/account-abstraction'
 
// Register a credential (ie. passkey).
const credential = await createWebAuthnCredential({ 
  name: 'Example', 
}) 
 
// Create a WebAuthn account from the credential.
const account = toWebAuthnAccount({
  credential,
})
```

On firefox it throws `NotAllowedError`, and on chrome it just hangs with a warning message:
```
publicKey.pubKeyCredParams is missing at least one of the default algorithm identifiers: ES256 and RS256. This can result in registration failures on incompatible authenticators. See https://chromium.googlesource.com/chromium/src/+/main/content/browser/webauth/pub_key_cred_params.md for details
```

**Comment by jxom** [comment]
@Silur – [example app](https://github.com/wevm/viem/tree/main/examples/smart-accounts_coinbase) that uses this snippet works for me: 

https://github.com/user-attachments/assets/49d17f2d-9d52-41f0-965a-f3911076120f

File: src/account-abstraction/clients/createPaymasterClient.ts [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+  transport extends Transport,
+  rpcSchema extends RpcSchema | undefined = undefined,
+>(
+  parameters: PaymasterClientConfig<transport, rpcSchema>,
+): PaymasterClient<transport, rpcSchema>
+
+export function createPaymasterClient(
+  parameters: PaymasterClientConfig,
+): PaymasterClient {
+  const { key = 'bundler', name = 'Bundler Client', transport } = parameters
```
Comment (holic) [inline src/account-abstraction/clients/createPaymasterClient.ts:69]: is this intentionally named like the bundler client rather than a paymaster client?

Comment (jxom) [reply src/account-abstraction/clients/createPaymasterClient.ts:69]: It should be "paymaster"

**Pull request merged**